lib/kcerts/ssh.go

```diff
@@ -0,0 +1,124 @@
+package kcerts
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"github.com/mitchellh/go-homedir"
+	"golang.org/x/crypto/ssh"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+const CAPrefix = "@cert-authority"
+const SSHDir = ".ssh"
+const KnownHosts = "known_hosts"
+
+// FindSSHDir will find the users ssh directory based on $HOME. If $HOME/.ssh does not exist
+// it will attempt to create it.
+func FindSSHDir() (string, error) {
+	hDir, err := homedir.Dir()
+	if err != nil {
+		return "", fmt.Errorf("could not find the home directory: %w", err)
+	}
+	sshDir := filepath.Join(hDir, SSHDir)
+	if err := os.Mkdir(sshDir, 0700); err != nil && !os.IsExist(err) {
+		return "", fmt.Errorf("could not create directory %s: %w", sshDir, err)
+	}
+	return sshDir, nil
+}
+
+// AddSSHCAToClient adds a public key to the $HOME/.ssh/known_hosts in the ssh-cert x509.1 format.
+// For each entry, it adds an additional line and does not concatenate
+func AddSSHCAToClient(publicKey ssh.PublicKey, hosts []string, sshDir string) error {
+	caPublic := string(ssh.MarshalAuthorizedKey(publicKey))
+	knownHosts := filepath.Join(sshDir, KnownHosts)
+	knownHostsFile, err := os.OpenFile(knownHosts, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("could not create known_hosts file: %w", err)
+		}
+		return err
+	}
+	existingKnownHostsContent, err := ioutil.ReadAll(knownHostsFile)
+	if err != nil {
+		return fmt.Errorf("error reading %s: %w", knownHosts, err)
+	}
+	defer knownHostsFile.Close()
```

Comment:
this should be moved on line 49, otherwise we will leak the file descriptor if ReadAll afterward fails.

Reply:
nice catch, done